### ISSUE TITLE
raidboss: trigger `override` proposal

### ISSUE THREAD

**Issue by quisquous** [quisquous/cactbot#1786]
I feel like there's been a lot of people asking in discord about how to override triggers lately, and so I've been trying to think about how to do this more cleanly.  cactbot already has (too) many ways to override trigger behavior (overriding conditions/text OR disabling+adding new triggers).

I feel like it's pretty hard to explain to people how to do these things.  Fundamentally, editing triggers is going to end up being "write some valid JavaScript" and I don't think there's value in some entirely new thing that's "easier" to write triggers in that's confusing for everybody.

I think one way to make this easier is to combine this overriding behavior into one step? For example, being able to copy and paste a trigger from a trigger file into your user/raidboss.js and then modifying what you wanted there.

I'm proposing `override: true` as a new field in a trigger.  If set to true, then any previously specified triggers with that id are fully disabled (not just hidden).  If you have multiple triggers with the same id and with `override: true` the last one specified will win and all previous will be disabled.  I think it's important to fully disable and not just hide, so that people copying and pasting full triggers will not generate double side effects from any JavaScript code.

(I think `override: 'id'` is a possibility too, but gets awkward if you have multiple user triggers overriding the same id? Then they both would run? At least with `override: true` the new trigger must have the same id as the old trigger, and so if you've copied and pasted multiple people's approaches to the same trigger, then only one will apply?)

(I guess another approach would be to have a TriggerOverrides field on Options instead of just Triggers and implicitly anything in there overrides the same thing with the same id, but I'm not sure I like that any better.  It would be a cleaner copy and paste, but it would also be more implicit.)

I could also imagine `override: true`, but on the trigger set entirely if somebody wants to copy and paste the whole thing? I think there would need to be some logic around handling the timeline file path here.

I think mostly I'm bringing this up because I feel like my previous overriding approaches have been super hacky.  I think I would prefer to get rid of the "condition / text overriding" if possible, as I think this `override: true` approach would allow you to do all of that more cleanly.  However, given the detritus of different approaches that haven't really worked out, I wanted to float this idea and see if anybody had criticisms, opinions, better ideas to make it easier to change text / conditions of triggers.

**Comment by JLGarber** [comment]
I think my only concern here is the same one we discussed previously, that the user might override a trigger that has side effects as well as returning popup text. Granted, if they're into the weeds of editing a trigger beyond "let's change the popup text", they *probably* should have at least a bit of expectation that Here Be Dragons.

Other than that (probably not perfectly solvable) issue, I think you're correct on your handling suggestion. I would suggest that we might want the override parameter to default true rather than false, however. Based on peoples' expectations in the ACT Discord server, I feel like that's the behavior most are expecting. Principle of Least Astonishment etc.

I very specifically like the idea of being able to hand users something to copypasta, since a number of the users I know personally are not terribly familiar with navigating text editors at all. Telling them "look for the [WHATEVER] keyword, find the brace next to it, insert a line below, and copy [THIS] there" could be significantly easier. Obviously that's not an ideal situation, since now they have potential cruft in their user file, but it is at least usable.

**Comment by quisquous** [comment]
> Other than that (probably not perfectly solvable) issue, I think you're correct on your handling suggestion. I would suggest that we might want the override parameter to default true rather than false, however. Based on peoples' expectations in the ACT Discord server, I feel like that's the behavior most are expecting. Principle of Least Astonishment etc.

Thanks, I was a little bit reluctant to go this implicit route, but I think it's the best.

I think the biggest negative to defaulting the override parameter to true is that there may be some people who have copypasta'd in the past and have duplicate trigger ids that aren't correct.  In the sense that they copied trigger A, but made new trigger B but used id A, and so now trigger A is now disabled unexpectedly.  So, this will be a breaking change.

I think it's ultimately the right decision, though.  The more I think about it, the more I am unsure what the use case of `override: false` is.  The only one I can think of is preventing future id collisions (i.e. if another trigger gets added with the same id in the future, then this one won't be disabled).  It also prevents your triggers from being overridden.  For example if a dev distributes an additional set of triggers that people can add to their user folder, then `override: false` would prevent them from being overridden, in the logical equivalent of `final` for triggers.  It's arguable whether that's a valid use case.

In summary, I think the plan here is that duplicate ids will automatically override any previously loaded trigger with that id.  It will  emit a log entry (maybe only in debug raidboss mode? maybe always??) when a trigger is overloaded.  There will be no `override` field, and this will all just be implicit behavior for triggers.  I think maybe for user folders we'll load in alphabetical order, so folks can control the order and overriding of triggers.
